A36-xds-for-servers.md

```diff
@@ -299,36 +299,94 @@ reference" and "release a reference" on the shared instance, or similar
+
+/** Destroys a config fetcher. */
+GRPCAPI void grpc_server_config_fetcher_destroy(
+    grpc_server_config_fetcher* config_fetcher);
 
-C will need to expose an API for C++ to utilize. We expect there will be an xDS
-filter that will be injected into the server. But we do not focus on that here
-and leave that as an implementation detail to be part of a future gRFC. Since C
-is monolithic and we are fine with xDS using internal APIs, there is quite a bit
-more flexibility in design options available for C than Java and Go.
-
-Create an `XdsServerBuilder` that mirrors the `ServerBuilder` API. It may be
-possible to implement it via delegating to a `ServerBuilder` instance, but that
-is an implementation detail.
-
-The `XdsServerBuilder` will need to use a C core API (e.g., a channel arg to
-pass a plugin instance) to enable xDS support in the built `Server`. The
-`XdsServerBuilder` can indirectly create the `XdsClient` and plumb it to filters
-or other code that may need it. Notably, it will _not_ be passed to the
-`ServerCredential`; the `ServerCredential` will be passed the configuration to
-use for each connection, like `ChannelCredential`, so won’t need to use the
-`XdsClient` for watches on the ADS stream. Since many `Server` methods are not
-`virtual`, the `Server` will _not_ be wrapped to add xDS functionality. This
-means the C server will need to be responsible for shutting down the
-`XdsClient`.
-
-Since `ServerBuilder` has few virtual methods, the `XdsServerBuilder` will not
-be directly interchangeable. We believe users will be minimally impacted by
-needing to refer to the precise type, especially with the availability of
-templates. If this becomes a problem more methods could become `virtual`.
+/** Sets the server's config fetcher.  Takes ownership. Must be called before 
+    adding ports. */
+GRPCAPI void grpc_server_set_config_fetcher(
+    grpc_server* server, grpc_server_config_fetcher* config_fetcher);
+```
+
+The server needs to be configured with the config fetcher before we add ports to
+the server. Additionally, when using a server config fetcher, the bound port
+might not be available immediately as a return value of the
+`grpc_server_add_secure_http2_port()` or the
+`grpc_server_add_insecure_http2_port()` API when using a wildcard port input.
+This arises from implementation details where Core currently invokes both
+`bind()` and `listen()` as part of `grpc_tcp_server_add_port()`, and as detailed
+earlier, since we would ideally not want to invoke `listen()` until we have a
+valid xDS configuration, `grpc_tcp_server_add_port()` is only invoked when the
+server is first ready to serve. (Note that any future transitions to the not
+serving state are still dealt with the `accept()`+`close()` method.) This
+behavior might change in the future by splitting `grpc_tcp_server_add_port` so
+that `bind()` and `listen()` are done separately, allowing `bind()` to be
```

Comment:
I thought we decided that the wildcard port was chosen by listen(). So this isn't applicable. I removed the language about splitting bind+listen long before merging the gRFC. The choices are "not listen()" or "accept()+close()".

Reply:
This is the relevant comment from @markdroth in the original PR
```
I think it's worth pointing out that calling listen() before the xDS resource is obtained is actually worse for clients, because even without calling accept(), it will still cause timeout delays before pick_first can fail over to the next address in the list. So I think implementations should prefer to defer calling listen() until after the xDS resource is obtained. That's what we're planning to do in C-core.

Also, the language here seems to imply that bind() should be done immediately even if listen() is delayed. Is there actually any reason for that? Based on what @yashykt and I are looking at in C-core, it seems like the only case where it would matter that bind() happens before we get the xDS resource is a server using an ephemeral port. And that seems like such a niche case that I think I'd be okay saying that we just don't support it for the xDS-enabled server API. (To be clear, it does seem better to call bind() earlier if possible, so that we can support ephemeral ports, but I think it's a nice-to-have, not a hard requirement. In C-core, it looks like that may be hard to de-couple, so at least initially we will probably start by deferring both bind() and listen() until after we have the xDS resource. We can revisit that later if someone actually wants to use ephemeral ports with xDS.)
```
From this I'm interpreting that splitting "bind()" and "listen()" is still a possible strategy.

Reply:
There was a lot more discussion than that[1][2][3], but in the end all the reference to bind() was removed in favor of focusing on listen() and "XdsServer will not support ephemeral ports as part of this gRFC but may be enhanced in the future."

I guess this text is fine as-is. It initially seemed like a lot about a topic that wasn't part of the gRFC, but it is really only one mention of splitting bind()+listen(), but another mention to bind() to make that make sense. That's fair. Let's keep it as-is.

1. https://github.com/grpc/proposal/pull/214#issuecomment-749280531
2. https://github.com/grpc/proposal/pull/214#discussion_r552172048
3. after other discussions, https://github.com/grpc/proposal/pull/214/commits/a933f33729a327fc153274b71fdd6a41b340b25b